cpp/src/binning/bin.cu

```diff
@@ -0,0 +1,222 @@
+        return (m_right_comp(value, m_right_span[index])) ? index : NULL_VALUE;
+    }
+
+    device_span<T> m_left_span{};  // The range of data containing all left bin edges.
+    device_span<T> m_right_span{};  // The range of data containing all right bin edges.
+    // TODO: Can I implement these as static members rather than making an instance on construction?
+    StrictWeakOrderingLeft m_left_comp{}; // Comparator used for left edges.
+    StrictWeakOrderingRight m_right_comp{}; // Comparator used for left edges.
+};
+
+
+struct filter_null_sentinel
+{
+    __device__ bool operator()(size_type i)
+    {
+        return (i != NULL_VALUE);
+    }
+};
+
+/// Bin the input by the edges in left_edges and right_edges.
+template <typename T, typename StrictWeakOrderingLeft, typename StrictWeakOrderingRight, bool InputIsNullable>
+std::unique_ptr<column> bin(column_view const& input, 
+                            column_view const& left_edges,
+                            column_view const& right_edges,
+                            rmm::mr::device_memory_resource * mr)
+{
+    auto output = cudf::make_numeric_column(data_type(type_to_id<size_type>()), input.size());
+    // TODO: Figure out why auto doesn't work here (or rather, why this works
+    // but then calls to begin/end later fail). I imagine suitable addition of
+    // a "template" keyword will fix it.
+    cudf::mutable_column_view output_mutable_view = output->mutable_view();
+
+    thrust::transform(thrust::device,
+            column_device_view::create(input)->pair_begin<T, InputIsNullable>(),
+            column_device_view::create(input)->pair_end<T, InputIsNullable>(),
+            output_mutable_view.begin<size_type>(),
+            // Must specify const T as the template type because the column
+            // views provided on the edges will always return const types. The
+            // template arguments to `datq` need not specify const since that
+            // const is added as part of the signature.
+            bin_finder<const T, StrictWeakOrderingLeft, StrictWeakOrderingRight>(
+                cudf::detail::device_span<const T>(left_edges.data<T>(), left_edges.size()),
+                cudf::detail::device_span<const T>(right_edges.data<T>(), right_edges.size())
+                )
+            );
+
+    auto mask_and_count = cudf::detail::valid_if(
+        output_mutable_view.begin<size_type>(),
+        output_mutable_view.end<size_type>(),
+        filter_null_sentinel()
+        );
+
+    output->set_null_mask(mask_and_count.first, mask_and_count.second);
+    return output;
+}
+
+template <typename T>
+constexpr inline auto is_supported_bin_type()
+{
+  // TODO: Determine what other types (such as fixed point numbers) should be
+  // supported, and whether any of them (like strings) require special
+  // handling.
+  return (cudf::is_numeric<T>() && not std::is_same<T, bool>::value); // || cudf::is_fixed_point<T>();
```

Comment:
Please enforce consistency for boolean operators. Don't mix `and`, `or`, `not` with `&&`.